coreos-cloudinit.go

```diff
@@ -285,6 +286,14 @@ func determineHostname(md datasource.Metadata, udata *initialize.UserData) strin
 			hostname = udataHostname
 		}
 	}
+	if len(hostname) > 63 {
```

Comment:
There was a discussion at some point about using the FQDN vs the short name as the hostname (and potentially adding the fqdn in `/etc/hosts`). Some projects (like afterburn) prefer to use the FQDN, some projects prefer the short name.

I think that if the hostname is too large, we should make an attempt to use the short name. After which we should truncate the short name if it continues to be too large.

The reason is that this way, we at least get a valid hostname, not a partial FQDN, if the short name is within the 63 character limit. The full FQDN will still be valid if set in `/etc/hosts`, if each element of the hostname is itself valid (63 characters after each dot, plus the rest of the rules of course).

@jepio @RiRa12621 what do you folks think?

Reply:
For example, if we have something like this:

```yaml
#cloud-config
hostname: this-hostname-is-larger-than-sixty-three-characters.example.com
```

And we truncate at 63 using the logic here, we get: `this-hostname-is-larger-than-sixty-three-characters`. Which is fine.

But if we have:

```yaml
#cloud-config
hostname: this-hostname-is-a-bit-larger-than-sixty-three-characters.example.com
```

we end up with: `this-hostname-is-a-bit-larger-than-sixty-three-characters.examp`. Which is neither the FQDN that we want, nor a short name. If we go with the short name approach, we at least end up with a valid short name of `this-hostname-is-a-bit-larger-than-sixty-three-characters`, and we can set the correct FQDN in `/etc/hosts`.

Reply:
For the second example, the code does the split on `.` before truncating, so the outcome is exactly as you describe (we don't end up with a partial FQDN).

I think in general the hostname and (/etc/hostname) should be just the first label, and the FQDN should be resolved through DNS.

cloud-init and coreos-cloudinit support `manage_etc_hosts: true` for those who need it.